File: android/app/src/main/java/com/noop/analytics/LevelBaselineStore.kt
package com.noop.analytics

import android.content.Context
import org.json.JSONObject

// MARK: - Keeping the yardstick still
//
// The baselines are derived once and then persisted verbatim. Every later read returns exactly what
// was written, so the level means the same thing in December as it did in March.
//
// WRITTEN ONCE, DELIBERATELY. [resolve] derives only when nothing is stored; it never re-derives, even
// when far more history is now available. A caller that wants a fresh scale has to say so through
// [refreeze], which is a visible, explicit act — the wearer should know their yardstick moved, because
// every level they remember was measured against the old one.
//
// STORED AS NUMBERS, NOT AS A SNAPSHOT DATE. What matters is the scale itself; when it was taken is
// recorded alongside only so the app can tell the wearer how thin the history behind it was.

object LevelBaselineStore {

    private const val KEY = "level.baselines"
    private const val KEY_FROZEN_AT = "level.baselinesFrozenAt"

    /**
     * The frozen baselines, deriving and storing them on the first call.
     *
     * [history] is only consulted when there is nothing stored, so the (potentially expensive) read of
     * a whole metric history can be a lambda the caller never pays for on a warm start.
     */
    fun resolve(
        context: Context,
        history: () -> Map<LevelMetric, List<Double>>,
    ): Map<LevelMetric, Baseline> {
        read(context)?.let { return it }
        val derived = LevelBaselines.deriveAll(history())
        write(context, derived)
        return derived
    }

    /** The stored set, or null when the scale has never been frozen. */
    fun read(context: Context): Map<LevelMetric, Baseline>? {
        val raw = prefs(context).getString(KEY, null) ?: return null
        return runCatching { decode(raw) }.getOrNull()?.takeIf { it.size == LevelMetric.entries.size }
    }

    /** When the scale was frozen, as epoch milliseconds, or null when it never was. */
    fun frozenAtMs(context: Context): Long? =
        prefs(context).getLong(KEY_FROZEN_AT, 0L).takeIf { it > 0L }

    /**
     * Derive the scale again from current history, replacing what was stored.
     *
     * The ONLY way the yardstick moves. Every level the wearer has seen was measured against the old
     * scale, so a caller must treat this as a visible change and not as maintenance.
     */
    fun refreeze(
        context: Context,
        history: Map<LevelMetric, List<Double>>,
    ): Map<LevelMetric, Baseline> {
        val derived = LevelBaselines.deriveAll(history)
        write(context, derived)
        return derived
    }

    private fun write(context: Context, baselines: Map<LevelMetric, Baseline>) {
        prefs(context).edit()
            .putString(KEY, encode(baselines))
            .putLong(KEY_FROZEN_AT, System.currentTimeMillis())
            .apply()
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences("noop_prefs", Context.MODE_PRIVATE)

    internal fun encode(baselines: Map<LevelMetric, Baseline>): String {
        val root = JSONObject()
        baselines.forEach { (metric, b) ->
            root.put(
                metric.name,
                JSONObject()
                    .put("mean", b.mean)
                    .put("sd", b.sd)
                    .put("min", b.min)
                    .put("max", b.max),
            )
        }
        return root.toString()
    }

    /**
     * Read a stored set back.
     *
     * A metric missing from the JSON falls back to the table rather than being dropped — a set with a
     * hole in it would silently stop scoring that component, and the wearer would see their level move
     * for no reason they could observe.
     */
    internal fun decode(raw: String): Map<LevelMetric, Baseline> {
        val root = JSONObject(raw)
        return LevelMetric.entries.associateWith { metric ->
            val o = root.optJSONObject(metric.name) ?: return@associateWith LevelBaselines.DEFAULT.getValue(metric)
            val mean = o.optDouble("mean", Double.NaN)
            val sd = o.optDouble("sd", Double.NaN)
            if (!mean.isFinite() || !sd.isFinite()) return@associateWith LevelBaselines.DEFAULT.getValue(metric)
            Baseline(
                mean = mean,
                sd = sd,
                min = o.optDouble("min", mean - 2 * sd),
                max = o.optDouble("max", mean + 2 * sd),
            )
        }
    }
}

File: Strand/Screens/LevelBaselineStore.swift
import Foundation
import StrandAnalytics

// LevelBaselineStore.swift — keeping the yardstick still.
//
// The baselines are derived from the wearer's own history and then persisted verbatim, so the level
// means the same thing in December as it did in March.
//
// FROZEN PER METRIC, ONCE EACH METRIC IS WORTH FREEZING. A metric with fewer than two weeks of readings
// is scored against the table for now and NOT written down; the first load on which it has enough of
// its own history freezes it. That matters more than it used to: daytime calm, for instance, only
// starts being recorded when this build first runs, and freezing it on day one would measure it against
// a stranger forever.
//
// v3: the level was rebuilt with no ceiling and new inputs (see `LevelEngine`); every older scale was
// for metrics that no longer exist.

enum LevelBaselineStore {

    private static let key = "level.baselines.v3"
    private static let frozenAtKey = "level.baselinesFrozenAt.v3"

    private struct Stored: Codable {
        let mean: Double
        let sd: Double
        let min: Double
        let max: Double
    }

    /// The baselines: every frozen metric as stored, every other one derived from `history` now — and
    /// frozen on the spot if its history has become deep enough.
    static func resolve(history: () -> [LevelMetric: [Double]]) -> [LevelMetric: Baseline] {
        var stored = readStored()
        var out: [LevelMetric: Baseline] = [:]
        var missing: [LevelMetric] = []
        for metric in LevelMetric.allCases {
            if let s = stored[metric.rawValue] {
                out[metric] = Baseline(mean: s.mean, sd: s.sd, min: s.min, max: s.max)
            } else {
                missing.append(metric)
            }
        }
        guard !missing.isEmpty else { return out }

        let readings = history()
        var changed = false
        for metric in missing {
            let h = readings[metric] ?? []
            let b = LevelBaselines.derive(metric, history: h)
            out[metric] = b
            if LevelBaselines.isDerivable(h) {
                stored[metric.rawValue] = Stored(mean: b.mean, sd: b.sd, min: b.min, max: b.max)
                changed = true
            }
        }
        if changed { write(stored) }
        return out
    }

    /// Whether every metric's scale is frozen.
    static var isFrozen: Bool { readStored().count == LevelMetric.allCases.count }

    static var frozenAt: Date? {
        let t = UserDefaults.standard.double(forKey: frozenAtKey)
        return t > 0 ? Date(timeIntervalSince1970: t) : nil
    }

    /// Derive every metric again from current history, replacing what was stored. The ONLY way a frozen
    /// yardstick moves.
    @discardableResult
    static func refreeze(history: [LevelMetric: [Double]]) -> [LevelMetric: Baseline] {
        var stored: [String: Stored] = [:]
        var out: [LevelMetric: Baseline] = [:]
        for metric in LevelMetric.allCases {
            let h = history[metric] ?? []
            let b = LevelBaselines.derive(metric, history: h)
            out[metric] = b
            if LevelBaselines.isDerivable(h) {
                stored[metric.rawValue] = Stored(mean: b.mean, sd: b.sd, min: b.min, max: b.max)
            }
        }
        write(stored)
        return out
    }

    private static func readStored() -> [String: Stored] {
        guard let data = UserDefaults.standard.data(forKey: key),
              let decoded = try? JSONDecoder().decode([String: Stored].self, from: data)
        else { return [:] }
        return decoded
    }

    private static func write(_ stored: [String: Stored]) {
        guard let data = try? JSONEncoder().encode(stored) else { return }
        UserDefaults.standard.set(data, forKey: key)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: frozenAtKey)
    }
}
